### ISSUE TITLE
Post triggers null reference exception in framework with no details as to why

### ISSUE THREAD

**Issue by TehWardy** [dotnet/aspnetcore#13853]
This was triggered by endpoint when posting an entity to an OData controller (recovered fro m the server console as the response produce was the json below.

**Response** 
```
{
	"error": "Object reference not set to an instance of an object."
}
```

**Console Logged Exception**

```
Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware:Error: An unhandled exception has occurred while executing the request.

System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
   at Microsoft.AspNetCore.Mvc.Internal.DefaultCollectionValidationStrategy.Enumerator.MoveNext()
   at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitChildren(IValidationStrategy strategy)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitComplexType(IValidationStrategy defaultStrategy)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.Visit(ModelMetadata metadata, String key, Object model)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitChildren(IValidationStrategy strategy)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitComplexType(IValidationStrategy defaultStrategy)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.Visit(ModelMetadata metadata, String key, Object model)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.Validate(ModelMetadata metadata, String key, Object model, Boolean alwaysValidateAtTopLevel)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ObjectModelValidator.Validate(ActionContext actionContext, ValidationStateDictionary validationState, String prefix, Object model, ModelMetadata metadata)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.EnforceBindRequiredAndValidate(ObjectModelValidator baseObjectValidator, ActionContext actionContext, ParameterDescriptor parameter, ModelMetadata metadata, ModelBindingContext modelBindingContext, ModelBindingResult modelBindingResult)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Builder.Extensions.MapWhenMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.InvokeCore(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context)

```

Any ideas how I debug this?

I lost 2 days to this on a production server due to the framework giving me literally no information what so ever (unless you consider the json response usefull) except the log from the VS console (which obviously can't be reported on a production box).

**Comment by TehWardy** [comment]
Having rolled back the EF relationship i was trying define the issue appears to be gone.
I would still consider this a bug in aspnet core as the information was vague and the exception raised seemingly unrelated to the issue that was actually caused.

Not sure how to reproduce this as a stand alone sample though.

**Comment by mkArtakMSFT** [comment]
@ajcvickers do you know what may have caused this?

**Comment by bricelam** [comment]
@smitpatel?

**Comment by bricelam** [comment]
Wait, that stack trace doesn't have EF in it...

**Comment by smitpatel** [comment]
At best it has `SelectManySingleSelectorIterator` which may be come through EF when doing client eval. But then we don't support client eval in that construct anymore in 3.0 so not sure if there is anything to react in EF till we get more info on the cause.

**Comment by TehWardy** [comment]
@bricelam lol that was my first reaction ... "eh, wtf? How did I get here!!!"

The other thing is it's not getting to the point of executing the controller action, it's like there was something about the construction of the EF model or something that triggered a failure "earlier in the request lifecycle".

Weirdly I managed to get this too ...
https://github.com/aspnet/AspNetCore/issues/14143

... I think this might be OData rather than AspNet Core directly but in situations where Badrequest(ModelState) is hit I think it could pull better information from the model state than it currently does which avoid the need for something like this in some situations at least ...
https://github.com/aspnet/AspNetCore/issues/13170 


... lots of referencing / linking there but hopefully this is useful information you guys can use to put some complex interaction scenarios together that are triggering some odd behaviour in the built in middleware or something.

I don't even mind if it's my code at fault, the key is in figuring out why my code is at fault in production environments (so no VS console). 

I've not looked at the source (assuming it's open source) but is there not a global handler that can return an aggregate exception to a user defined handler so I can at least globally say "ok AspNet Core drop the actual problem to this log file and return a 500 response with maybe the exception message in it".

To be honest though, what i needed from EF is essentially an unsupported scenario so maybe I hit some sort of edge case in the framework(s) here?

**Comment by davidfowl** [comment]
I highly doubt the issue is in your code, it's likely a bug in this implementation of our enumerator impl https://github.com/dotnet/aspnetcore/blob/940dcb0aba247da054ee435380d533e04944093c/src/Mvc/Mvc.Core/src/ModelBinding/Validation/DefaultCollectionValidationStrategy.cs#L88. 

What would help is more data about the type of model that caused this.

**Comment by TehWardy** [comment]
Given the age of this issue I have likely put some hack workaround in place to avoid this some time ago.
I have found however that the frameworks understanding of relationships is a bit peculiar and it's behaviour regarding data I know might be structurally correct but maybe not meet the validation requirements is frustrating.

So lets say I have a complex object that looks something like this ...

```
public class Invoice
{
      public string SupplierRef { get set; }
      public string BuyerRef {get; set; }
}
```

... then I build something like this ...

```
public class ValidationResult<T>
{
      public T Source {get; set; }
     public bool IsValid { get; set; }
     public string Message { get; set; }
}
```

I might have on my endpoint a method that is named something like "ValidateReferences"
that I can pass an Invoice to and it could tell me if the Invoice has values that correctly point to a known buyer / supplier in the system.

In this situation, lets say my client code is parsing data from a third party when I validate using business logic if I add validation errors to the model state or produce something that isn't considered "metadata valid" in the controller code as my result the serialisation just simply stops and the connection is dropped at the property value that fails leading to half serialised responses.

It seems that the Odata framework will refuse to allow serialisation of invalid data.
And I started to notice this sort of weirdness in both directions.

I have requested somewhere (but can't seem to find the ticket at the moment) the ability to either disable for a particular action or mark an action as having something about it such that my business logic will deal with this issue so I don't want the framework to intercept it as a "general rule" that I could apply to any action.

I've also got situations where I have interim objects that are somewhere populated but not fully complete, known to be in this state but need to check something with the server.
This presents a problem as partial entity posts hit this sort of pitfall.
It could be argued that I should create a method that handles the specifics but in my third party data situation I don't know how correct or incorrect the data is and I need the server to check that for me so I just send what I have.

I admit though that there's probably some precedent here for perhaps a smarter client design ... I often just ask for the thing with that reference (for example) as a means to check if it's there instead but when you start considering this problem over batches of objects it gets call heavy so hence why I reverted to the "let the back end figure it out" approach as that could tell me in SQL the few issues that my posted values had by slicing the object up and asking various questions on the business layer.
